cudax/include/cuda/experimental/__launch/confiuration.cuh

```diff
@@ -0,0 +1,474 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDAX__LAUNCH_CONFIGURATION
+#define _CUDAX__LAUNCH_CONFIGURATION
+#include <cuda/experimental/hierarchy.cuh>
+#include <cuda/std/span>
+#include <cuda/std/tuple>
+
+#if _CCCL_STD_VER >= 2017
+namespace cuda::experimental
+{
+
+template <typename Dimensions, typename... Options>
+struct kernel_config;
+
+namespace detail
+{
+struct launch_option
+{
+  static constexpr bool needs_attribute_space = false;
+  static constexpr bool is_relevant_on_device = false;
+
+protected:
+  _CCCL_NODISCARD cudaError_t apply(cudaLaunchConfig_t& config, void* kernel) const noexcept
+  {
+    return cudaSuccess;
+  }
+};
+
+template <typename Dimensions, typename... Options>
+cudaError_t apply_kernel_config(
+  const kernel_config<Dimensions, Options...>& config, cudaLaunchConfig_t& cuda_config, void* kernel) noexcept;
+
+// Might need to go to the main namespace?
+enum class launch_option_kind
+{
+  cooperative_launch,
+  dynamic_shared_memory,
+  launch_priority
+};
+
+struct option_not_found
+{};
+
+template <detail::launch_option_kind Kind>
+struct find_option_in_tuple_impl
+{
+  template <typename Option, typename... Options>
+  _CCCL_DEVICE auto& operator()(const Option& opt, const Options&... rest)
+  {
+    if constexpr (Option::kind == Kind)
+    {
+      return opt;
+    }
+    else
+    {
+      return (*this)(rest...);
+    }
+  }
+
+  _CCCL_DEVICE auto operator()()
+  {
+    return option_not_found();
+  }
+};
+
+template <detail::launch_option_kind Kind, typename... Options>
+_CCCL_DEVICE auto& find_option_in_tuple(const ::cuda::std::tuple<Options...>& tuple)
+{
+  return ::cuda::std::apply(find_option_in_tuple_impl<Kind>(), tuple);
+}
+
+template <typename...>
+inline constexpr bool no_duplicate_options = true;
+
+template <typename Option, typename... Rest>
+inline constexpr bool no_duplicate_options<Option, Rest...> =
+  ((Option::kind != Rest::kind) && ...) && no_duplicate_options<Rest...>;
+
+} // namespace detail
+
+/**
+ * @brief Launch option enabling cooperative launch
+ *
+ * This launch option causes the launched grid to be restricted to a number of
+ * blocks that can simultaneously execute on the device. It means that every thread
+ * in the launched grid can eventualy observe execution of each other thread in the grid.
+ * It also enables usage of cooperative_groups::grid_group::sync() function, that
+ * synchronizes all threads in the grid.
+ *
+ * @par Snippet
+ * @code
+ * #include <cudax/launch.cuh>
+ * #include <cooperative_groups.h>
+ *
+ * template <typename Configuration>
+ * __global__ void kernel(Configuration conf)
+ * {
+ *     auto grid = cooperative_groups::this_grid();
+ *     grid.sync();
+ * }
+ *
+ * void kernel_launch(cuda::stream_ref stream) {
+ *     auto dims = cudax::make_hierarchy(cudax::block<128>(), cudax::grid(4));
+ *     auto conf = cudax::make_configuration(dims, cooperative_launch());
```

Comment:
Question: Since C++17 (which is old by today's standards), CTAD is preferred over make funcitons if possible. Why wouldn't CTAD work in this case?

Reply:
For the hierarchy, I wanted a separate function, so you can list the levels both ascending or descending and it will be correctly sorted, I am not sure if that is possible with CTAD, maybe it is.
For the configuration I see less reasons to have a separate function, so there is a good chance it can be just cudax::kernel_config constructor